src/aviary/functional.py

```diff
@@ -0,0 +1,175 @@
+from collections.abc import Callable
+from functools import update_wrapper
+from typing import Any
+
+from pydantic import BaseModel, Field
+
+from aviary.env import Environment, Frame
+from aviary.message import Message
+from aviary.tools import Messages, Tool, ToolRequestMessage
+from aviary.utils import is_coroutine_callable
+
+
+class DynamicState(BaseModel):
+    """Dynamic env state model that adapts to provided extras."""
+
+    reward: float = 0
+    done: bool = False
+    extras: dict[str, Any] = Field(default_factory=dict)
+
+    def __getattr__(self, name: str):
+        """Allow direct access to extras as attributes."""
+        if name in self.extras:
+            return self.extras[name]
+        raise AttributeError(f"'State' has no attribute {name!r}")
+
+
+class FunctionalEnvironment(Environment[DynamicState]):
+    """Environment class for function-based environments."""
```

Comment:
Can you add a comment here saying that tools for this environment need to encapsulate `done` logic via `DynamicState`

Reply:
The complete usage is in the `start` docstring - I added a note to see that.